File: 22-generate_parentheses.py
# time: O(2^n)
# space: O(2^n) - recursive stack
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def dfs(opening, closing, s):
            nonlocal combos
            # base case
            # append to results if the size is right
            # rest of the code guarantees well-formed string
            if len(s) == n * 2:
                combos.append(s)
                return 

            # add opening parenthese to string up until we reach n (0-indexed)
            if opening < n:
                dfs(opening + 1, closing, s + "(")

            # add closing parenthese to string up until we reach n (0-indexed)
            if closing < opening:
                dfs(opening, closing + 1, s + ")")

        combos = []
        dfs(0, 0, "")

        return combos

File: 22-generate_parentheses.cs
public class Solution {
    public IList<string> GenerateParenthesis(int n)
    {
        List<string> ans = new();
        dfs(0, 0, "");
        return ans;

        void dfs(int open, int close, string s)
        {
            if (s.Length == 2 * n)
            {
                ans.Add(s);
                return;
            }

            if (open < n)
            {
                dfs(open + 1, close, ")" + s);
            }
            if (close < open)
            {
                dfs(open, close + 1, "(" + s);
            }
        }
    }
}
